Mk strict version check

mkangia opened the pull request:
Just some possible scope of improvement i saw during my work on https://github.com/dimagi/commcare-hq/pull/20757  regarding different formats for version we have in HQ code base.
I checked with the mobile team the format they would usually keep for version and Clayton said its always `major.minor.point`. Though over HQ code base at places 
1. we have skipped point version or 
2. added some text with it like `1.2.3/latest` or `1.2.latest`.
Looking at code [here](https://github.com/dimagi/commcare-hq/blob/049b1f45fcdcb88d79d4804728055f514a51dfa9/corehq/apps/builds/models.py#L216) its clear that `point` should be always preset which confirms # 1.
Looking at the way we mention version on [commcare builds](https://www.commcarehq.org/builds/edit_menu/) its clear the `point` is always an integer and i think that the BuildSpec would always have version from CommCare Builds added to the environment.

This PR does two things
1. update check to also ensure point version i.e 1.2 is not valid, 1.2.dev is not valid, 1.2.1 is.
2. add check on BuildSpec as well. This is quite a strict check. I don't expect this to cause any issue on the live environment since all commcare builds have been added in correct format but it would cause issue locally if we have created builds with invalid format like i had 1.2.dev

On staging:
Removed the builds with invalid version numbers

On prod:
Renamed 1.2.dev -> 1.2.999, 1.2.0alpha -> 1.2.998, 1.3.0alpha -> 1.3.999, 2.25.w -> 2.25.999

On India:
All good.

On PNA:
All good.

On Swiss:
All good.

On ICDS-CAS:
All good.

mkangia committed:
strict check for point version as well

mkangia committed:
add check on Build Spec as well

mkangia committed:
fix version

mkangia committed:
fix version

mkangia committed:
fix version to allow expected features

mkangia committed:
fix version

mkangia committed:
fix version

esoergel commented on corehq/apps/app_manager/tests/test_media_suite.py:
```diff
@@ -152,7 +152,7 @@ def setUp(self):
         self.module = self.app.add_module(Module.new_module("Module 1", None))
         self.form = self.app.new_form(0, "Form 1", None)
-        self.min_spec = BuildSpec.from_string('2.21/latest')
+        self.min_spec = BuildSpec.from_string('2.21.0/latest')
```
This one still has `latest` in it

mkangia replied:
`from_string` parses the version and build_number(which is latest) here.

esoergel commented on corehq/apps/app_manager/tests/test_case_list_form.py:
```diff
@@ -404,7 +404,7 @@ def test_target_module_different_datums(self):
             './entry')
 
     def test_case_list_form_requires_parent_case_but_target_doesnt(self):
-        factory = AppFactory(build_version='2.9.0/latest')
```
Do you know what this `/latest` thing is?  It sounds like "the latest build, but at least 2.9.0" Could this exist in the real world too?

mkangia replied:
No seeing an instance when version is set with `latest` in it on [HQ](https://www.commcarehq.org/builds/edit_menu/). 
Probably just a quick way of setting the version and the latest flag. I don't see how its set from the HQ interface though.

esoergel replied:
It might not come from the UI.  Here's a bit of code that sets that version format:
https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/app_manager/models.py#L4837-L4837
I'm not sure what it does - just found that from grepping around.  There's also this constant that may factor in somewhere:
https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/app_manager/models.py#L191-L191
Anyways, I don't know how these are used, but changing the code to not accept that format seems risky.

@orangejenny Do you know anything about this version format?

mkangia replied:
Looking at the first link there is a comment there
> # scrape for old conventions and get rid of them

Looks like this was used earlier and then there was this set up to migrate that from old convention where it was stored on Application to new convention where its now on BuildSpec?

also pinging @sravfeyn for hints.. looks like he did some work on this in https://github.com/dimagi/commcare-hq/pull/17614/files

sravfeyn replied:
The word 'latest' in https://github.com/dimagi/commcare-hq/pull/17614/ is not really a string that would be appended to a build version (so, not related to this issue). It's just an HTML form value to keep track of a certain setting of the app.

mkangia replied:
@esoergel 
after some digging found this 
https://github.com/dimagi/commcare-hq/commit/bd56ca7d3aea35925013e458c710e68f8139437d
blast from the past.

snopoke replied:
If you look at the code here for [BuildSpec.from_string](https://github.com/dimagi/commcare-hq/blob/049b1f45fcdcb88d79d4804728055f514a51dfa9/corehq/apps/builds/models.py#L208) you can see that `latest` means the latest build for this version.

Looking at the [full list of versions of CommCare](https://www.commcarehq.org/builds/) you'll notice that there are multiple builds with the same version number.

So we can't just remove `latest`.

snopoke replied:
Offlined about this but updating here as well. `AppFactor` passes the version directly into the build spec so the `/latest` part doesn't get parsed. So removing it from these places makes sense.

dannyroberts commented:
Hey @mkangia, can you update the description with a little more context regarding the motivation behind this change (fogbugz ticket, roadmap item, or just a description of why this came up)? That would help me understand the intention and reason about whether it'll do that.

In general, I would be pretty careful about changing the semantics of this build version stuff, because it's one of the earliest parts of commcare-hq and hasn't been touched in a long time, so it's not very well documented and the effects of changing it may not be that widely understood.

mkangia commented:
Sure @dannyroberts ,
Added more context to description

>it's one of the earliest parts of commcare-hq and hasn't been touched in a long time

Yup!

snopoke commented on corehq/apps/builds/models.py:
```diff
@@ -166,7 +169,7 @@ def j2me_enabled_build_versions(cls):
 
 class BuildSpec(DocumentSchema):
-    version = StringProperty()
+    version = SemanticVersionProperty(required=False)
```
This seems OK but why make this not required?

mkangia replied:
When you add a new application object, the version is blank so this is to avoid failure there.

snopoke commented on corehq/apps/builds/views.py:
```diff
@@ -127,7 +127,7 @@ def import_build(request):
 
     try:
-        SemanticVersionProperty().validate(version)
+        SemanticVersionProperty(required=True).validate(version)
```
`required=True` is the default so this change isn't necessary

mkangia replied:
Had to add this because of [this](https://github.com/dimagi/jsonobject/blob/010391db9daabde26051511cd974f23677504f93/jsonobject/base_properties.pyx#L110).
The value that is passed during `__init__` is different from what is passed to validate and is relevant.

snopoke commented on corehq/apps/builds/management/commands/commcare_build_importer.py:
```diff
@@ -101,8 +101,8 @@ def add_build(self, builf_file, version, build_number):
-            {352167: 'http://jenkins.dimagi.com/job/commcare-mobile-2.21/352167/',
-             352366: 'http://jenkins.dimagi.com/job/commcare-mobile-2.21/352366/'}
+            {352167: 'http://jenkins.dimagi.com/job/commcare-mobile-2.21.0/352167/',
+             352366: 'http://jenkins.dimagi.com/job/commcare-mobile-2.21.0/352366/'}
```
Did you look at Jenkins? If you had you may have noticed that none of the jobs include the `point` in the name. The reason being that the same build job will be used for all point releases of that version.

mkangia replied:
Didn't refer Jenkins. Was just looking at how we mention versions on HQ. Would this change effect things on jenkins?

mkangia replied:
to be reverted.

snopoke commented on corehq/apps/app_manager/suite_xml/xml_models.py:
```diff
@@ -681,13 +681,13 @@ class ActionMixin(OrderedXmlObject):
 
 class LocalizedAction(ActionMixin, TextOrDisplay):
-    """ For CC >= 2.21 """
+    """ For CC >= 2.21.0 """
```
Specifying the `point` isn't necessary for any of these since we're only interested in when the feature was released (unless there's a specific case where we need to restrict it to after a bug was fixed)

mkangia replied:
Ya, its just to bring the structure change. 2.21 should be 2.21.0

esoergel replied:
This is minor, but I think 2.21 better corresponds to the mental model - conceptually this is a feature introduced in 2.21, not 2.21.0.  Technically yes, that was the first release in the minor version, but that's not the important part.  When talking about feature availability, it's only major and minor versions - think Django docs that say "Introduced in Django 1.10".

If you use `distutils.version.StrictVersion`, they're the same anyways:
```python
In [2]: StrictVersion('2.21') == StrictVersion('2.21.0')
Out[2]: True
```

esoergel replied:
We might be stuck with `LooseVersion` to deal with the `".../Latest"` type business.  Either way, `LooseVersion('2.21') is always considered less than `LooseVersion('2.21.anything')`.  I think it makes sense to allow for omitting the the point release when checking a minimum version, but that anything declaring a specific app build should specify the point release.  I don't think it matters much one way or the other though, and I'm fine with this implementation.

mkangia replied:
Cool, thanks @esoergel.

snopoke replied:
I agree that it doesn't make a huge difference but my preference is also for the shorter version since they are semantically the same.

snopoke reviewed (commented):
I don't think your observations and assumptions are correct:

> Looking at code [here](https://github.com/dimagi/commcare-hq/blob/049b1f45fcdcb88d79d4804728055f514a51dfa9/corehq/apps/builds/models.py#L216) its clear that point should be always preset which confirms # 1.

The code you link to completely ignores the `point` part of the release. Understanding how release versions work this makes sense since the `point` part indicates a PATCH release which is essentially a bug fix. So it would make sense that HQ would ignore that for the most part. More info on Semantic Versioning here: https://semver.org/


Most of this code was related to building JAR files for old Java phones. I think the only thing we need this for now is enabling / disabling features based on selected CommCare version.

At this point I don't think it's worth a big effort to make that change since this code doesn't require much maintenance and isn't very complex.

mkangia commented:
@snopoke  thanks for the review

> The code you link to completely ignores the point part of the release.
My Understanding is that [here](https://github.com/dimagi/commcare-hq/blob/049b1f45fcdcb88d79d4804728055f514a51dfa9/corehq/apps/builds/models.py#L216) if the version is something like '2.1' it would raise an exception
```
In [2]: major, minor, _ = version.split('.')
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-2-544e585ddd09> in <module>()
----> 1 major, minor, _ = version.split('.')

ValueError: need more than 2 values to unpack
```

>If you look at the code here for BuildSpec.from_string you can see that latest means the latest build for this version.

`build_version` is directly assigned as the `version` for the BuildSpec in case of AppFactory and the `from_string` would also split `2.9.0/latest` into `2.9.0` as the version and if it has latest passed within the argument it marks the boolean `latest` on the BuildSpec. hence the format for `version` would still end up being without `latest`

snopoke reviewed (approved):
Offlined with @mkangia and went through this in more detail. I think these changes are fine (apart from the changes to the comments related to Jenkins job names).

mkangia committed:
revert change to version on jenkins links

mkangia commented:
Going to hold off on merge and merge this on Monday.

dannyroberts commented:
I'm not sure whether these changes are fine, but some of the description gave me the impression that our ideas of what the original code is doing differ.

For example, in the current code,

```
self._require_minimum_version('2.2')
```
means "this thing requires anything >= 2.2", which includes "2.2.0", "2.2.1", etc.—it's not saying it believes there's some version out there that's actually called "2.2". Because of how [semantic versioning](https://semver.org/) is supposed to work (and that the mobile releases follow it), using '2.2' there instead of '2.2.0' (which has the same meaning) has the nice quality that it reinforces the notion that mobile _features_ (which for the most part are what restrict the use of HQ features) change with minor release usually, and changing with a point/hotfix/patch release (the third number) is uncommon. So in this situation, I think the old code is doing a good job.

There are also lots of places where you change something like `1.2.3/latest` to `1.2.3`. I also want to point out that there, it's not that the code thinks that there is some version that is called "1.2.3/latest"—the "/latest" thing is something that other HQ code interprets as a signifier to get the latest build that is tagged as "1.2.3"—if I recall correctly, conceptually this is so that when the mobile team is testing out the latest unreleased version, and they push different builds all tagged as the same version, the latest of them (the one with the highest build number) is picked. See also @snopoke's comment https://github.com/dimagi/commcare-hq/pull/20845/files#r196328320 along the same lines, which I agree with.

If these ideas differ from the mental model used in making this change, I'd say we should step back and rethink it. My impression is that these two things were being interpreted differently in the course of making this PR.

mkangia commented:
@dannyroberts 
Thanks for the review..

> So in this situation, I think the old code is doing a good job.

Ya, i think i am fine with keeping this `a.b` if that makes more sense but it might cause some issue as mentioned [here](https://github.com/dimagi/commcare-hq/pull/20845#discussion_r196512657)

> There are also lots of places where you change something like 1.2.3/latest to 1.2.3

That convention should still works since `1.2.3/latest` is parsed by [from_string](https://github.com/dimagi/commcare-hq/blob/f892bbd292358b0331844581460a4f4e4f926029/corehq/apps/builds/models.py#L211) method which i have not changed. I have added limitation on the `version` property of BuildSpec. @snopoke and I offlined about his comment and cleared the confusion about this change.

mkangia committed:
Revert "fix version"

This reverts commit fe659065718b2e4bc054464191c94f17923681fe.

mkangia committed:
revert changes to minimum versions

snopoke commented on corehq/apps/builds/management/commands/commcare_build_importer.py:
```diff
@@ -29,7 +29,7 @@ def interactive_import(self):
         logging.getLogger("requests").setLevel(logging.WARNING)
 
-        # let user chose a project from the list ['commcare-mobile-2.21', ...]
+        # let user chose a project from the list ['commcare-mobile-2.21.0', ...]
```
this is still referring to a jenkins project name so the 'point' should be removed

mkangia replied:
thought that was just an example so didnt revert it.. Fine either ways so reverted now.

mkangia committed:
revert change to jenkin version options

dannyroberts commented on corehq/apps/builds/models.py:
```diff
@@ -16,10 +16,13 @@ class SemanticVersionProperty(StringProperty):
+            major, minor, point = value.split('.')
             int(major)
             int(minor)
+            int(point)
```
I had a though here, which is that this might interfere with legitimate version conventions for non-released versions (2.25.0a for alpha, 2.25.0b for beta, 2.25.0rc for release candidate), but if we're not using any of those right now, then I suppose we don't have to worry about it right now. Happy to make it strict now if current usage supports that, and then make it lax in the future if it's needed.

dannyroberts reviewed (approved):
@mkangia Thanks for the patience and the explanations here. It's helped me understand the changes more accurately, and I'm happy with this change now!

mkangia commented:
Putting this on staging for now.
